### DISCUSSION TITLE
What is the best way to specify that a matrix variable in cvxpy is lower triangular?

### DISCUSSION THREAD

**Discussion by abyvinod** [cvxpy/cvxpy#2515] [General]
I would like to optimize over the set of lower triangular matrices. Currently, I do the following in CVXPY:
```
import cvxpy as cp

n = 5
L = cp.Variable((n,n))
const = []
for row_index in range(n - 1):
    const += [L[row_index, row_index + 1 :] == 0]
# Rest of the code adds more constraints to const, and then optimizes...
```
This is wasteful (in memory and possibly computation?) since I am defining quite a few unnecessary variables.

With YALMIP (MATLAB), I could define a lower triangular matrix as follows:
```
L_symm = sdpvar(n, n);      % Define symmetric matrix and then focus on its lower half
L_ltri = tril(L_symm);    
```
Is there an equivalent approach in cvxpy?

**Comment by Transurgeon** [comment]
Hello @abyvinod we currently are working towards adding support for sparse variables in cvxpy. 
This could potentially include allowing to only optimise over lower triangular matrices.

**Comment by rileyjmurray** [comment]
@abyvinod for now your best bet is as follows:

Form a (sparse) matrix whose columns are vectorized versions of matrices
with a one at position (i,j) for i <= j and zero elsewhere. Call this
matrix T_basis.

Compute dim=int(np.sqrt(T.shape[0]))

Define a vector variable t_vec of length dim*(dim+1)/2.

Set L = (T_basis @ t_vec).reshape((dim, dim))

A workflow sort of like this should do what you need.
